### PR TITLE
JENKINS-39977: Add support for Organization Folder items and GitHub a…

### PR DESCRIPTION
**Pull request by mkobit**
…nd Bitbucket SCM navigators

- GitHub Branch source support
- Bitbucket Branch source support

Left out of this iteration:

- Folder icon
- Pipeline libraries
- Health metrics.
  There is only 1 available metric implementation now that is not very useful to the organization folder item type.
- Git SCM navigator

### CODE REVIEW

**Comment by mkobit** [comment]
It is possible/probable that some of this code can be probably be stripped down to reuse some of the stuff for the `multibranchPipelineJob`, but I wanted to get some feedback first. It may also be able to use some of the generated Job DSL for the organization types but I didn't know the best way to test that.

**Comment by daspilker** [comment]
Yes, strip down this PR and use the Automatically Generated DSL where possible.

File: job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/workflow/OrganizationFolderTriggerContext.groovy [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  OrganizationFolderTriggerContext(JobManagement jobManagement, Item item) {
+    super(jobManagement, item)
+  }
+
+  /**
+   * Sets the periodic trigger for when this has not been triggered.
+   * @since 1.56
+   */
+  void periodicIfNotOtherwiseTriggered(PeriodicFolderTrigger periodicFolderTrigger) {
+    triggerNodes << new NodeBuilder().'com.cloudbees.hudson.plugins.folder.computed.PeriodicFolderTrigger' {
```
Comment (mkobit) [inline job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/workflow/OrganizationFolderTriggerContext.groovy]: Open question: the [`MultibranchWorkflowTriggerContext`](https://github.com/jenkinsci/job-dsl-plugin/blob/master/job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/triggers/MultibranchWorkflowTriggerContext.groovy#L18-L23) has a different pattern, but the UI presents a dropdown here. What is the best option to go with?

At least from my standpoint, we switched from `Multibranch` to this SCM Navigator style almost entirely, so consistency may matter for the Job DSL.

File: job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/DslFactory.groovy [hunk 2]
```diff
@@ -155,6 +156,23 @@ interface DslFactory extends ViewFactory {
     MultibranchWorkflowJob multibranchPipelineJob(String name)
 
     /**
+     * Creates or updates an organization folder job.
+     *
+     * @since 1.56
+     * @see #organizationFolderJob(java.lang.String)
+     */
+    @RequiresPlugin(id = 'branch-api', minimumVersion = '1.11')
+    OrganizationFolderJob organizationFolderJob(String name, @DslContext(OrganizationFolderJob) Closure closure)
```
Comment (mkobit) [inline job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/DslFactory.groovy]: Open question: `organizationFolderJob` or `organizationFolder`?

**Comment by mkobit** [comment]
@daspilker I rebased off of `master` and took another crack at this by allowing the generated DSL for the actual navigators.

I added a couple open questions I had to the code, would love some feedback.

File: build.gradle [hunk 3]
```diff
@@ -13,6 +13,7 @@ buildscript {
 plugins {
     // need to be in the root project as workaround for https://github.com/michel-kraemer/gradle-download-task/issues/48
     id 'de.undercouch.download' version '3.1.2'
+    id 'build-dashboard'
```
Comment (daspilker) [inline build.gradle]: move this to a separate pull request if useful

**Comment by daspilker** [comment]
I added several comments, but this is going in the right direction.

**Comment by mkobit** [comment]
@daspilker I only see one comment, you may have to publish the review

**Comment by mkobit** [comment]
Sorry I haven't gotten back to this, will hopefully get back to this PR this weekend but if not definitely next week.

**Comment by mkobit** [comment]
@daspilker , made some changes based on your comments so I think this could be ready for another round of review!

**Comment by daspilker** [comment]
Looks good! I squashed the changes and merged them manually.

**Comment by sgtcoolguy** [comment]
Forgive me for bugging you on this, but I don't see these commits in the master branch @daspilker - did this ever get merged in/pushed? I'm dying to get a new release to use this since I'm in the process of moving over all my jobs to pipeline Jenkinsfile defined jobs - and don't like that right now the github org scans that coordinate them all aren't in my DSL/SCM...

**Comment by daspilker** [comment]
Arg, I forgot to push, sorry. Try again. I'm planning to cut a release end of next week.
